nomad/core_sched_test.go

```diff
@@ -2224,3 +2193,113 @@ func TestAllocation_GCEligible(t *testing.T) {
+	node.CSINodePlugins = map[string]*structs.CSIInfo{
+		"csi-plugin-example": {PluginID: "csi-plugin-example",
+			Healthy:  true,
+			NodeInfo: &structs.CSINodeInfo{},
+		},
+	}
+	err := state.UpsertNode(99, node)
+	require.NoError(t, err)
+	volId0 := uuid.Generate()
+	vols := []*structs.CSIVolume{{
+		ID:             volId0,
+		Namespace:      "notTheNamespace",
+		PluginID:       "csi-plugin-example",
+		AccessMode:     structs.CSIVolumeAccessModeMultiNodeSingleWriter,
+		AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+	}}
+	err = state.CSIVolumeRegister(100, vols)
+	require.NoError(t, err)
+	vol, err := state.CSIVolumeByID(ws, volId0)
+	require.NoError(t, err)
+	require.Len(t, vol.ReadAllocs, 0)
+	require.Len(t, vol.WriteAllocs, 0)
+
+	// Create a job with 2 allocations
+	job := mock.Job()
+	job.TaskGroups[0].Volumes = map[string]*structs.VolumeRequest{
+		"_": {
+			Name:     "someVolume",
+			Type:     structs.VolumeTypeCSI,
+			Source:   volId0,
+			ReadOnly: false,
+		},
+	}
+	err = state.UpsertJob(101, job)
+	require.NoError(t, err)
+
+	alloc1 := mock.Alloc()
+	alloc1.JobID = job.ID
+	alloc1.NodeID = node.ID
+	err = state.UpsertJobSummary(102, mock.JobSummary(alloc1.JobID))
+	require.NoError(t, err)
+	alloc1.TaskGroup = job.TaskGroups[0].Name
+
+	alloc2 := mock.Alloc()
+	alloc2.JobID = job.ID
+	alloc2.NodeID = node.ID
+	err = state.UpsertJobSummary(103, mock.JobSummary(alloc2.JobID))
+	require.NoError(t, err)
+	alloc2.TaskGroup = job.TaskGroups[0].Name
+
+	err = state.UpsertAllocs(104, []*structs.Allocation{alloc1, alloc2})
+	require.NoError(t, err)
+
+	// Claim the volumes and verify the claims were set
+	err = state.CSIVolumeClaim(105, volId0, alloc1, structs.CSIVolumeClaimWrite)
+	require.NoError(t, err)
+	err = state.CSIVolumeClaim(106, volId0, alloc2, structs.CSIVolumeClaimRead)
+	require.NoError(t, err)
+	vol, err = state.CSIVolumeByID(ws, volId0)
+	require.NoError(t, err)
+	require.Len(t, vol.ReadAllocs, 1)
+	require.Len(t, vol.WriteAllocs, 1)
+
+	// Update the 1st alloc as failed/terminated
+	alloc1.ClientStatus = structs.AllocClientStatusFailed
+	err = state.UpdateAllocsFromClient(107, []*structs.Allocation{alloc1})
+	require.NoError(t, err)
+
+	// Create the GC eval we'd get from Node.UpdateAlloc
+	now := time.Now().UTC()
+	eval := &structs.Evaluation{
+		ID:          uuid.Generate(),
+		Namespace:   job.Namespace,
+		Priority:    structs.CoreJobPriority,
+		Type:        structs.JobTypeCore,
+		TriggeredBy: structs.EvalTriggerAllocStop,
+		JobID:       structs.CoreJobCSIVolumeClaimGC + ":" + job.ID,
+		LeaderACL:   srv.getLeaderAcl(),
+		Status:      structs.EvalStatusPending,
+		CreateTime:  now.UTC().UnixNano(),
+		ModifyTime:  now.UTC().UnixNano(),
+	}
+
+	// Process the eval
+	snap, err := state.Snapshot()
+	require.NoError(t, err)
+	core := NewCoreScheduler(srv, snap)
+	err = core.Process(eval)
+	require.NoError(t, err)
+
+	// Verify the claim was released
```

Comment:
@langmartin wanted to flag this for you in particular because you implemented the state store work. As far as I can tell there's no real reason to bother with tracking `PastAllocs` with this implementation, and we can swap out the implementation of `structs.CSIVolume.ClaimRelease` with that of `structs.CSIVolume.GCAlloc`. Do you have any thoughts here?

Reply:
> PastAllocs at this point is there to serve as a debugging tool. If there's a user sensible gap of several minutes between the reapVolumes that would detect the ReadAlloc + Terminal state and the eval reap that deletes the allocations, it's worth keeping both stages.

I was trying to wrap up the RFC section to explain this, but `PastAllocs` as it stands right now doesn't get us anything. When we get `Node.UpdateAlloc` for a terminal alloc, we can't move that alloc out of `Read/WriteAllocs` because that would make it eligible for scheduling before we've released the claim. At the very least we'd need to have a `PastReadAllocs` and a `PastWriteAllocs` _and_ check that during scheduling, but right now we're not looking in `PastAllocs` at all during scheduling. 

Alternately, we could add the alloc to `PastAllocs` but _not_ remove it from `ReadAllocs/WriteAllocs` until it's GC'd, but I'm not sure that helps us in any way that isn't better served by checking if the alloc is terminal.

Reply:
I think the way you have it working is correct. Here's how I think it should be go:

1. `UpdateAlloc` informs the state store that the alloc is terminal
2. The volume and eval are marked for GC
3. If it's unclaimed, we `ControllerUnpublish`
4. we `ClaimRelease`, and move the alloc to `PastAllocs`
5. `EvalGCThreshold` later, the the eval is garbage collected, and on alloc GC we delete it from `PastAllocs`

The gap between 4 & 5 is the user perceivable bit that `PastAllocs` gets us, I think. I'm a bit fuzzy on the details of where the eval or job is marked for GC to actually reap the eval (and allocs) but I think that configured duration kicks in either way.

Reply:
The eval to GC the volume is immediately queued-up on `Node.UpdateAlloc`. We don't wait until after `EvalGCThreshold`. Otherwise the volume would be unavailable for claiming for up to several minutes. There is of course a short gap between the time the eval is queued and it's dequeued by the broker, but unless there's anything consuming `PastAllocs`, recording that doesn't get us anything.

We _also_ run the volume GC as part of job GC to make sure that we don't have any stragglers if the controller fails on us, but that's a cleanup, not the primary way of releasing volume claims.

Reply:
Ok, I think I get what you're proposing at least.

But when we GC the job/alloc we end up having to run the volume GC process anyways, because we don't have any guarantees they're not running concurrently (we can interleave transactions with the lengthy `ControllerUnpublish`). So the `PastAllocs` _might_ be useful but could just as easily be removed instantly depending on timing, which makes for an unreliable debugging instrument that we have to pay for in extra raft transactions.

Reply:
Also, without anything currently consuming `PastAllocs` I'm feeling extra-skeptical about its use at this stage of the design.

Reply:
Ok, that's fair. We can always re-introduce it if we need to.